### PR TITLE
V0.4.0

### PR DESCRIPTION
**Pull request by kneelsdev**

### CODE REVIEW

File: contracts/ERC721A__OwnableUpgradeable.sol [hunk 1]
```diff
@@ -45,10 +45,14 @@ abstract contract ERC721A__OwnableUpgradeable is ERC721A__Initializable, ERC721A
      * @dev Throws if called by any account other than the owner.
      */
     modifier onlyOwner() {
-        require(owner() == _msgSenderERC721A(), "Ownable: caller is not the owner");
+        _isOwner();
         _;
     }
 
+    function _isOwner() internal view {
```
Comment (kneelsdev) [inline contracts/ERC721A__OwnableUpgradeable.sol:52]: Doing this saves bytecode
https://soliditydeveloper.com/max-contract-size

File: contracts/Archetype.sol [hunk 2]
```diff
@@ -111,26 +133,19 @@ contract Archetype is ERC721A__Initializable, ERC721AUpgradeable, OperatorFilter
   // VARIABLES
   //
   mapping(bytes32 => Invite) public invites;
-  mapping(address => mapping(bytes32 => uint256)) private minted;
-  mapping(address => uint128) public affiliateBalance;
-  mapping(uint256 => bytes) private tokenMsg;
+  mapping(address => mapping(bytes32 => uint256)) private _minted;
```
Comment (coindegen) [inline contracts/Archetype.sol:136]: can we make this public? 
https://github.com/scatter-art/contracts/issues/22

File: contracts/Archetype.sol [hunk 3]
```diff
@@ -653,24 +727,24 @@ contract Archetype is ERC721A__Initializable, ERC721AUpgradeable, OperatorFilter
   }
 
   modifier onlyPlatform() {
-    require(PLATFORM == _msgSenderERC721A(), "caller is not the platform");
+    require(PLATFORM == msg.sender, "caller is not the platform");
```
Comment (coindegen) [inline contracts/Archetype.sol:730]: should we use a custom Error for this?

**Comment by coindegen** [comment]
I'm getting following errors when compiling:

```
 contracts git:(v0.4.0) ✗ npx hardhat compile
TypeError: Derived contract must override function "_isConstructor". Two or more base classes define function with same name and parameter types.
  --> contracts/Archetype.sol:49:1:
   |
49 | contract Archetype is
   | ^ (Relevant source part starts here and spans across multiple lines).
Note: Definition in "Initializable": 
  --> @openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol:77:5:
   |
77 |     function _isConstructor() private view returns (bool) {
   |     ^ (Relevant source part starts here and spans across multiple lines).
Note: Definition in "ERC721A__Initializable": 
  --> erc721a-upgradeable/contracts/ERC721A__Initializable.sol:62:5:
   |
62 |     function _isConstructor() private view returns (bool) {
   |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 2.
   --> contracts/Archetype.sol:762:5:
    |
762 |     onlyAllowedOperatorApproval(operator)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 2.
   --> contracts/Archetype.sol:770:5:
    |
770 |     onlyAllowedOperatorApproval(operator)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 2.
   --> contracts/Archetype.sol:779:21:
    |
779 |   ) public override onlyAllowedOperator(from) {
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^


TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 2.
   --> contracts/Archetype.sol:787:21:
    |
787 |   ) public override onlyAllowedOperator(from) {
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^


TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 2.
   --> contracts/Archetype.sol:796:21:
    |
796 |   ) public override onlyAllowedOperator(from) {
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^


Error HH600: Compilation failed
```

**Comment by coindegen** [comment]
Did you see these errors at all? 

```
TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:774:3:
    |
774 |   function approve(address operator, uint256 tokenId)
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:386:5:
    |
386 |     function approve(address to, uint256 tokenId) public payable virtual override {
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:774:3:
    |
774 |   function approve(address operator, uint256 tokenId)
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:386:5:
    |
386 |     function approve(address to, uint256 tokenId) public payable virtual override {
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:782:3:
    |
782 |   function transferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:493:5:
    |
493 |     function transferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:782:3:
    |
782 |   function transferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:493:5:
    |
493 |     function transferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:790:3:
    |
790 |   function safeTransferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:559:5:
    |
559 |     function safeTransferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:790:3:
    |
790 |   function safeTransferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:559:5:
    |
559 |     function safeTransferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:798:3:
    |
798 |   function safeTransferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:582:5:
    |
582 |     function safeTransferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Overriding function changes state mutability from "payable" to "nonpayable".
   --> contracts/Archetype.sol:798:3:
    |
798 |   function safeTransferFrom(
    |   ^ (Relevant source part starts here and spans across multiple lines).
Note: Overridden function is here:
   --> erc721a-upgradeable/contracts/ERC721AUpgradeable.sol:582:5:
    |
582 |     function safeTransferFrom(
    |     ^ (Relevant source part starts here and spans across multiple lines).


Error HH600: Compilation failed

For more info go to https://hardhat.org/HH600 or run Hardhat with --show-stack-traces
```

**Pull request merged**